### DISCUSSION TITLE
[Feature] PAR2 protection of super-important pictures & videos

### DISCUSSION THREAD

**Discussion by JarlMorgennebel** [immich-app/immich#2593] [Feature Request]
### The feature

There are videos and pictures which I cannot re-create - super-important ones.

My kids as toddlers. First day at school. Last day at school. The first friend. A wedding etc.

Protecting these memories against bitrot and/or defective disks is super-critical. The file system ZFS and btrfs do protect againt bitrot, but not everyone is using them or has the right hardware (ECC protected memory as example).

Would it be possible to support https://en.wikipedia.org/wiki/Parchive par2-files as additional protection for selected videos & pictures as an additional protection solution? par2 files could be created for files in an album and favorite files only.

The tools are available as Linux package. It's a matter of filtering and running a background job as well as ignoring par2 files within the picture library itself.

Thanks, -JMN

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by lorenz** [comment]
IMO this is not something for Immich to address. 

Hardware requirements: ZFS and BTRFS do not require ECC memory. It's just recommended as otherwise data corruption can occur in memory which cannot reasonably be protected against with a software-side checksum. But this problem doesn't just affect BTRFS and ZFS, but also every other filesystem and software (including Immich). So ECC memory would still be recommended, even if using Immich with PAR2.

Bitrot and modern storage: I run mid-size storage systems (in the order of 100s of TiB) and basically most failures you see on spinning disks are either corrupted blocks/sectors which happen relatively frequently (in the order of around 0.1/TiB/Year). These basically make 4KiB of data inaccessible but are otherwise unproblematic, your storage layer just has to recompute that data from parity and write it back. The disks still run fine even with sometimes hundreds of those over the years. The other one are disk faults. These generally manifest either as a rapidly-increasing number of bad blocks or the disk no longer responding to commands. These result in the loss of an entire disk unless you pay for expensive data recovery.

For SSDs bad blocks are more rare (but do still happen), instead you get controller firmware bugs which cause the controller to crash or refuse to start, rendering the data on the disk inaccessible. You also get firmware bugs which cause the disk to read complete garbage (which ZFS/btrfs will pick up and throw the disk out of the pool). Also SSDs die if written to too much.

So basically on modern storage bitflips are very rare and are mostly dealt with by the hardware itself. Memory bitflips happen (at a highly hardware-specific rate), but these require ECC memory to fix, PAR2 will not do that. It is much more likely that you are unable to read entire blocks because of corruption issues (which do get detected by the disk) or lose a disk. So basically if you care about your data, run a BTRFS or ZFS mirror or (for ZFS) a raidz. If you don't have the budget for ECC memory, throughly test your memory before with memtest86 or similar.

And ideally even then you should have a backup of the data off-site in cold storage.

**Reply by JarlMorgennebel** [reply]
I agree with you 100% - in a professional environment.

But who is the target audience of Immich - professionals (running 100TB of storage, have backup, restore the backup regulary and validate) or private home users who manage their memories on smaller and underpowered devices...?

I consider myself a professional, I do follow a 3-2-1 backup scheme, I do have an offsite encrypted backup. But I also lost data - single bit flips changing the colours, unreadable files, lost videos. But I am not sure when I lost them (I have 250.000+ pictures) because I did not checked them regulary.  A solution like par2 would help me in case of - but I also understand the argument that it is beyond Immich's scope and should be solved by the server instead.

par2 for me is a quick and dirty solution, as my data consolidates at Immich and needs to be protected there until I have time, money and learnt enough to build something with ZFS.

**Reply by lorenz** [reply]
What disks were you running on? IMO bit flips are most likely a memory issue, not a disk issue, at least with modern (2010+) disks.
